pkg/cache/listers_test.go

```diff
@@ -0,0 +1,155 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"testing"
+
+	"github.com/kcp-dev/logicalcluster"
+	"github.com/stretchr/testify/require"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/tools/cache"
+)
+
+func newTestIndexer() cache.Indexer {
+	indexer := cache.NewIndexer(
+		ClusterAwareKeyFunc,
+		cache.Indexers{
+			ClusterIndexName:             ClusterIndexFunc,
+			ClusterAndNamespaceIndexName: ClusterAndNamespaceIndexFunc,
+		})
+
+	for _, cluster := range []string{"c1", "c2"} {
+		indexer.Add(newUnstructured(cluster, "ns1", "n1", map[string]string{"app": "myapp"}))
+		indexer.Add(newUnstructured(cluster, "ns2", "n1", map[string]string{"app": "myapp"}))
+		indexer.Add(newUnstructured(cluster, "ns2", "n2", nil))
+
+		indexer.Add(newUnstructured(cluster, "", "cn1", map[string]string{"app": "myapp"}))
+		indexer.Add(newUnstructured(cluster, "", "cn2", nil))
+	}
+
+	return indexer
+}
+
+func TestGenericClusterLister(t *testing.T) {
+	indexer := newTestIndexer()
+
+	l := NewGenericClusterLister(indexer, schema.GroupResource{})
+
+	tests := map[string]struct {
+		selector   labels.Selector
+		desiredLen int
+	}{
+		"nil":        {selector: nil, desiredLen: 10},
+		"everything": {selector: labels.Everything(), desiredLen: 10},
+		"app: myapp": {selector: labels.Set{"app": "myapp"}.AsSelector(), desiredLen: 6},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			list, err := l.List(tt.selector)
+			require.NoError(t, err)
+			require.Len(t, list, tt.desiredLen)
+		})
+	}
+}
+
+func TestGenericLister(t *testing.T) {
+	indexer := newTestIndexer()
+
+	l := NewGenericClusterLister(indexer, schema.GroupResource{})
+
+	tests := map[string]struct {
+		cluster    string
+		selector   labels.Selector
+		desiredLen int
+		name       string
+	}{
+		"c1,nil":        {cluster: "c1", selector: nil, desiredLen: 5},
+		"c2,nil":        {cluster: "c2", selector: nil, desiredLen: 5},
+		"c1,everything": {cluster: "c1", selector: labels.Everything(), desiredLen: 5},
+		"c2,everything": {cluster: "c2", selector: labels.Everything(), desiredLen: 5},
+		"c1,app:myapp":  {cluster: "c1", selector: labels.Set{"app": "myapp"}.AsSelector(), desiredLen: 3},
+		"c2,app:myapp":  {cluster: "c2", selector: labels.Set{"app": "myapp"}.AsSelector(), desiredLen: 3},
+		"c1,cn1":        {cluster: "c1", name: "cn1"},
+		"c1,cn2":        {cluster: "c1", name: "cn2"},
+		"c2,cn1":        {cluster: "c2", name: "cn1"},
+		"c2,cn2":        {cluster: "c2", name: "cn2"},
+	}
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			lister := l.ByCluster(logicalcluster.New(tt.cluster))
+
+			if tt.name == "" {
+				list, err := lister.List(tt.selector)
+
+				require.NoError(t, err)
+				require.Len(t, list, tt.desiredLen)
```

Comment:
I kinda think this needs to be more explicit about what we're validating, i.e. that you're getting the right cluster name, and right labels.